File: quake-log-parser/parser/parser.go
package parser

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func ParseLog(filePath string) (map[string]*Game, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	games := make(map[string]*Game)
	gameID := 0
	currentGame := NewGame()

	killRegex := regexp.MustCompile(`(\d+):(\d+) Kill: (\d+) (\d+) (\d+): (.*) killed (.*) by (.*)`)

	for scanner.Scan() {
		line := scanner.Text()

		if strings.Contains(line, "InitGame") {
			gameID++
			currentGame = NewGame()
			games["game_"+strconv.Itoa(gameID)] = currentGame
			fmt.Printf("Initialized game %d\n", gameID)
		} else if strings.Contains(line, "Kill:") {
			matches := killRegex.FindStringSubmatch(line)
			if len(matches) > 0 {
				mod := matches[8]
				killer := matches[6]
				killed := matches[7]

				currentGame.TotalKills++
				fmt.Printf("Game %d: Kill event - Killer: %s, Killed: %s, Mod: %s\n", gameID, killer, killed, mod)

				if killer != "<world>" {
					currentGame.AddPlayer(killer)
				}
				currentGame.AddPlayer(killed)

				if killer != "<world>" {
					currentGame.addPoints(killer, 1)
					if killer == killed { // Suicide check
						currentGame.addPoints(killer, -1)
					}
				} else {
					currentGame.addPoints(killed, -1)
				}

				if _, exists := currentGame.KillsByMean[mod]; !exists {
					currentGame.KillsByMean[mod] = 0
				}
				currentGame.KillsByMean[mod]++
			}
		}
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return games, nil
}

func (g *Game) addPoints(player string, points int) {
	g.Points[player] += points
}

File: quake-log-parser/parser.js
const fs = require('fs');

class Game {
    constructor() {
        this.total_kills = 0;
        this.players = new Set();
        this.ranking = {};
        this.kills_by_mean = {};
    }

    addPlayer(player) {
        if (player && player !== '<world>') { // if the player is not '<world>' and is not null it tries to add a new player to the set of players, if it has not already been added it initializes the player's ranking with 0
            this.players.add(player);
            if (!this.ranking[player]) {
                this.ranking[player] = 0;
            }
        }
    }

    addKill(killer, killed, mod) {
        this.total_kills++; // increments the total number of kills

        this.addPlayer(killer); // tries adding the killer to the set of players
        this.addPlayer(killed); // tries adding the killed to the set of players

        if (killer !== '<world>') { // if the killer is not '<world>' it increments the killer's point by 1, if it is '<world>' it decrements the killed's point by 1
            this.ranking[killer]++; 
            if (killer === killed) { // if the killer is the same as the killed it decrements the killer's point by 1
                this.ranking[killer] -= 1; // Suicide
            }
        } else {
            this.ranking[killed]--;
        }

        this.kills_by_mean[mod] = (this.kills_by_mean[mod] || 0) + 1;
    }

    getPlayers() {
        return Array.from(this.players); // returns the set as a array
    }

    getSortedPoints() { // sorts the object ranking to return ranking 
        return Object.entries(this.ranking)
            .sort(([, a], [, b]) => b - a)
            .reduce((acc, [key, val]) => ({ ...acc, [key]: val }), {}); 
    }
}

const parseLog = (filePath) => { // reads the log file 
    const logContent = fs.readFileSync(filePath, 'utf-8');
    const lines = logContent.split('\n');
    const games = {};

    let currentGame = null;
    let gameID = 0;

    const killRegex = /(\d+):(\d+) Kill: (\d+) (\d+) (\d+): (.*) killed (.*) by (.*)/; // regex to find kill logs 

    lines.forEach(line => {
        if (line.includes('InitGame')) { // checks the line for InitGame, if this is another game, it makes sure to initiate new structures for it 
            gameID++;
            currentGame = new Game();
            games[`game_${gameID}`] = currentGame;
        } else if (line.includes('Kill:')) { // checks the line for Kill:, if it is a kill log, it extracts the killer, killed and mod from the line and adds the kill to the current game
            const matches = killRegex.exec(line);
            if (matches) {
                const [, , , , , , killer, killed, mod] = matches; // extracts the killer, killed and mod from the line. The killer is the 6th element, the killed is the 7th element and the mod is the 8th element.
                currentGame.addKill(killer, killed, mod); // adds the kill to the current game
            }
        }
    });

    return games;
};

module.exports = { parseLog, Game };
